lib/orbit/import/rfid.ex

```diff
@@ -20,39 +22,56 @@ defmodule Orbit.Import.Rfid do
 
   @spec import_rows(Enumerable.t()) :: :ok
   def import_rows(rows) do
-    imported_count =
+    rows =
       rows
       |> Enum.map(fn row ->
         Map.update!(row, @badge_number_field, &String.trim_leading(&1, "0"))
       end)
-      |> Enum.group_by(& &1[@badge_number_field])
-      |> Enum.reduce(
-        0,
-        fn {badge_number, badge_serials}, acc ->
-          count =
+
+    num_serials_with_duplicates =
+      rows
+      |> Enum.frequencies_by(& &1[@badge_serial_field])
+      |> Enum.count(fn {_serial, count} -> count > 1 end)
+
+    if num_serials_with_duplicates > 0 do
+      Logger.warning("rfid_import badge_serials_with_duplicates=#{num_serials_with_duplicates}")
+    end
+
+    serials_by_badge =
+      Enum.group_by(rows, & &1[@badge_number_field], & &1[@badge_serial_field])
+
+    {:ok, imported_count} =
+      Repo.transaction(fn ->
```

Comment:
**musing (non-blocking):** I'm sad that this means we can't just easily leverage built-in Ecto functionality and the `on_replace` argument to the `has_many` association, but given the nature of the constraint across badge serials for different employees I don't think there's really a better way to do it. In any event I think this transaction should do the trick of making the changes atomic, and I'm not too worried about race conditions because there isn't really other functionality writing this table concurrently (especially because the employee info and RFID imports are offset by half an hour).